src/contexts/DataProvider.js

```diff
@@ -0,0 +1,56 @@
+import React, {
+  createContext,
+  useCallback,
+  useState,
+  useContext,
+  useEffect,
+} from "react";
+import { GeoContext } from "../components/GeoProvider";
+
+export const DataContext = createContext(null);
+
+const DataProvider = ({ children }) => {
+  const [data, setData] = useState([]);
+  const [selectedLocation, setSelectedLocation] = useState("All");
+  const [locations, setLocations] = useState([]);
+  const { coords, isGeolocationAvailable, mode } = useContext(GeoContext);
+
+  const getUrl = useCallback(() => {
+    let url = `.netlify/functions/providers?location=${selectedLocation}`;
+
+    if ((mode === 'geo' && isGeolocationAvailable) || mode === 'postcode') {
+      if (coords) {
+        url = `${url}&coords=${coords.latitude},${coords.longitude}`;
+      }
+    }
+
+    return url;
+  }, [coords, isGeolocationAvailable, selectedLocation, mode]);
+
+  useEffect(() => {
+    const url = getUrl();
+
+    fetch(url)
+      .then((response) => response.json())
+      .then(async (data) => {
+        const [first, ...results] = data;
+        setData([first, ...results]);
+
+        const locationSet = new Set();
+        data.forEach((provider) => {
+          locationSet.add(provider["provider town/city"]);
+        });
+        setLocations(["All", ...locationSet]);
+      });
+  }, [selectedLocation, getUrl]);
+
+  return (
+    <DataContext.Provider
+      value={{ data, selectedLocation, setSelectedLocation, locations }}
+    >
```

Comment:
This cleans things up a lot, nice work! I'm thinking it may be worth adding in an isLoading and maybe error state to the fetch. This would allow us to very easily show a loading spinner/message and error messages.
I would suggest we keep the logic for showing the loading spinner within this provider (e.g. have an if(isLoading) just before the main return statement) rather than passing it down to each child and having duplicate logic for showing the spinner/message across the app. Let me know what you think! 

Also this lib looks like it could be pretty handy for just this sort of thing, but possibly overkill atm - https://github.com/vercel/swr

Reply:
agreed about loading state. It's actually already been added in #18, just haven't merged it in yet. swr looks good - if you wanted to branch off this and implement it, that'd be cool!

Reply:
loading state from #18 added to the DataProvider